sys/net/application_layer/ng_zep/ng_zep.c

```diff
@@ -0,0 +1,1067 @@
+
+    if (tmp == NULL) {
+        DEBUG("zep: Could not get write access to received packet\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    pkt = tmp;
+
+    while (pkt->next) {
+        /* remove everything below UDP */
+        ng_pktbuf_remove_snip(pkt, pkt->next);
+    }
+
+    hdr = pkt->data;
+
+    if ((pkt->size < 2) || (hdr->preamble[0] != 'E') ||
+        (hdr->preamble[1] != 'X')) {
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    switch (hdr->version) {
+        case 1:
+            pkt = _recv_v1(dev, pkt);
+            break;
+
+        case 2:
+            pkt = _recv_v2(dev, pkt);
+            break;
+
+        default:
+            ng_pktbuf_release(pkt);
+            return;
+    }
+
+    if (pkt != NULL && dev->event_cb != NULL) {
+        dev->event_cb(NETDEV_EVENT_RX_COMPLETE, pkt);
+    }
+    else if (pkt != NULL) {
+        ng_pktbuf_release(pkt);
+    }
+}
+
+/* TODO: Generalize and move all below to ng_ieee802154 */
+static size_t _make_data_frame_hdr(ng_zep_t *dev, uint8_t *buf,
+                                   ng_netif_hdr_t *hdr)
+{
+    int pos = 0;
+
+    /* we are building a data frame here */
+    buf[0] = NG_IEEE802154_FCF_TYPE_DATA;
+    buf[1] = 0x88;      /* use short src and dst addresses as starting point */
+
+    /* if AUTOACK is enabled, then we also expect ACKs for this packet */
+    if (dev->flags & NG_ZEP_FLAGS_AUTOACK) {
+        buf[0] |= NG_IEEE802154_FCF_ACK_REQ;
+    }
+
+    /* fill in destination PAN ID */
+    pos = 3;
+    buf[pos++] = dev->pan.u8[0];
+    buf[pos++] = dev->pan.u8[1];
+
+    /* fill in destination address */
+    if (hdr->flags &
+        (NG_NETIF_HDR_FLAGS_BROADCAST | NG_NETIF_HDR_FLAGS_MULTICAST)) {
+        buf[pos++] = 0xff;
+        buf[pos++] = 0xff;
+    }
+    else if (hdr->dst_l2addr_len == 2) {
+        uint8_t *dst_addr = ng_netif_hdr_get_dst_addr(hdr);
+        buf[pos++] = dst_addr[1];
+        buf[pos++] = dst_addr[0];
+    }
+    else if (hdr->dst_l2addr_len == 8) {
+        buf[1] |= 0x04;
+        uint8_t *dst_addr = ng_netif_hdr_get_dst_addr(hdr);
+
+        for (int i = 7;  i >= 0; i--) {
+            buf[pos++] = dst_addr[i];
+        }
+    }
+    else {
+        /* unsupported address length */
+        return 0;
+    }
+
+    /* fill in source PAN ID (if applicable) */
+    if (dev->flags & NG_ZEP_FLAGS_USE_SRC_PAN) {
+        buf[pos++] = dev->pan.u8[0];
+        buf[pos++] = dev->pan.u8[1];
+    }
+    else {
+        buf[0] |= NG_IEEE802154_FCF_PAN_COMP;
+    }
+
+    /* fill in source address */
+    if (dev->flags & NG_ZEP_FLAGS_SRC_ADDR_LONG) {
+        buf[1] |= 0x40;
+        memcpy(&(buf[pos]), &dev->eui64, 8);
+        pos += 8;
+    }
+    else {
+        memcpy(&(buf[pos]), &dev->addr, 2);
+        pos += 2;
+    }
+
+    /* set sequence number */
+    buf[2] = dev->seq++;
+    /* return actual header length */
+    return pos;
+}
+
+static size_t _get_frame_hdr_len(uint8_t *mhr)
```

Comment:
Me neither… As I said: I copied this from the then current version of Hauke's [`ng_at86rf2xx`](https://github.com/RIOT-OS/RIOT/blob/master/drivers/ng_at86rf2xx/ng_at86rf2xx_netdev.c#L105), since they generate the desired IEEE 802.15.4 header as far as I can tell.